src/open-uri.c

```diff
@@ -729,7 +733,7 @@ handle_open_in_thread_func (GTask *task,
         app = latest_id;
       else if (default_app != NULL)
         app = default_app;
-      else if (choices && app_exists (choices[0]))
+      else if (choices && choices[0] && app_exists (choices[0]))
```

Comment:
Can you make find_recommended_choices() return early if the `infos` list is empty instead of returning an array with just the NULL terminator?

Reply:
I could, but I'm not 100% sure I should. How strongly do you prefer this?

It has to set choices to *something*: either `NULL`, or a non-`NULL` pointer to an empty strv (just the `NULL` terminator). At the moment it looks as though it always sets choices to non-`NULL` (although possibly empty), which is the same behaviour as `g_variant_dup_strv()`.

I could make it early-return if you strongly prefer that, but that's actually more code and less consistent. We'd have to make sure it isn't passed to any of the GStrv functions that expect a non-`NULL` argument (like `g_strv_contains()`, `g_strv_equal()`, `g_strv_length()`), and that `xdp_impl_app_chooser_call_update_choices()` and `xdp_impl_app_chooser_call_choose_application()` will accept `NULL`.

If your objection is to the apparent redundancy here, I could sidestep it like this:

```suggestion
      else if (n_choices > 0 && app_exists (choices[0]))
```

Would you like that any better?

Reply:
> and that `xdp_impl_app_chooser_call_update_choices()` and `xdp_impl_app_chooser_call_choose_application()` will accept `NULL`

I don't think they do: they end up in GVariant format string `^as`, which calls `g_variant_new_strv (ptr, -1)`, which wants a non-`NULL` *ptr* (`NULL` is only allowed if the length is explicitly 0, not if it's -1).

Reply:
The diff would look something like this:

```diff
diff --git a/src/open-uri.c b/src/open-uri.c
index d503702..556fa71 100644
--- a/src/open-uri.c
+++ b/src/open-uri.c
@@ -509,6 +509,14 @@ find_recommended_choices (const char *scheme,
   if (!infos)
     infos = g_app_info_get_all_for_type (content_type);
 
+  /* Don't return a non-NULL array that is empty */
+  if (!infos)
+    {
+      *choices = NULL;
+      *choices_len = 0;
+      return;
+    }
+
   n_choices = g_list_length (infos);
   result = g_new (char *, n_choices + 1);
   for (l = infos, i = 0; l; l = l->next)
@@ -542,6 +550,7 @@ app_info_changed (GAppInfoMonitor *monitor,
   content_type = (const char *)g_object_get_data (G_OBJECT (request), "content-type");
   find_recommended_choices (scheme, content_type, &default_app, &choices, &n_choices);
 
+  /* TODO: make sure this accepts NULL for choices */
   xdp_impl_app_chooser_call_update_choices (impl,
                                             request->id,
                                             (const char * const *)choices,
@@ -733,7 +742,7 @@ handle_open_in_thread_func (GTask *task,
         app = latest_id;
       else if (default_app != NULL)
         app = default_app;
-      else if (choices && choices[0] && app_exists (choices[0]))
+      else if (choices && app_exists (choices[0]))
         app = choices[0];
 
       if (app)
@@ -786,6 +795,7 @@ handle_open_in_thread_func (GTask *task,
 
   g_debug ("Opening app chooser");
 
+  /* TODO: make sure this accepts NULL for choices */
   xdp_impl_app_chooser_call_choose_application (impl,
                                                 request->id,
                                                 app_id,
```

Reply:
I'm fine with either ways of fixing, but I think it should either always be non-NULL, with only a terminator, or it should be NULL when the array is empty. So your first suggestion is probably fine.

Reply:
> I think it should either always be non-NULL, with only a terminator, or it should be NULL when the array is empty

OK, that makes sense!